Fixed #31784 -- Fixed crash when sending emails on Python 3.6.11+, 3.7.8+, and 3.8.4+.

felixxm opened the pull request:
Maybe it's too naive :thinking: @apollo13 Can you take a look?


ticket-31784

apollo13 commented on tests/mail/tests.py:
```diff
@@ -756,19 +756,32 @@ def test_sanitize_address(self):
+            (
+                'Tó Example very long' * 4 + ' <to@example.com>',
+                'utf-8',
+                '=?utf-8?q?T=C3=B3_Example_very_longT=C3=B3_Example_very_longT=C3=B3_Example_very_longT=C3=B3_Example_very_long?= '
```
I am not exactly sure about that change of mine. It is probably okay that `sanitize_address` returns a long address, but I'll have to double check if actually serializing a message like that rewraps it again.

apollo13 commented on tests/mail/tests.py:
```diff
@@ -730,14 +730,14 @@ def test_sanitize_address(self):
                 ('A name', 'to@example.com'),
                 'utf-8',
-                '=?utf-8?q?A_name?= <to@example.com>',
+                'A name <to@example.com>',
```
There should never be a reason to encode ascii to something else. While technically valid it makes mails bigger (especially if at one point we decide to call sanitize address for all addresses)

ewjoachim replied:
I think the last time this code changed, we did this because it was the way it was done before, but I also support not encoding already correct ascii. Chances are no one will complain that this changed.

felixxm replied:
Maybe we should add a short note about this change in release notes :thinking:

apollo13 commented:
So previously we did `nm = Header(nm, encoding).encode()` on `nm` and then passed that into `Address`. Now we still do it but no longer pass it into `Address` and as such the new validation that checks for `\r & \n` is no longer triggered (You'd get those after `Header.encode` ran over a to long name).

carltongibson commented:
OK, I see. ... 

I need to have a little play. It looks like we're doing the exact [same `encode` two-step as `formataddr`](https://github.com/python/cpython/blob/015efdbef7454a522e88cd79ba2b4cd77a5fb2a2/Lib/email/utils.py#L92-L99).

(This is probably fine. No doubt you've already been though the various thoughts here.)

apollo13 commented:
> I need to have a little play. It looks like we're doing the exact same encode two-step as formataddr.

Ha, yeah I ran into that trap too. `charset.header_encode` != `header_encode_lines`. What `formataddr` misses to do is splitting after 76 characters -- this is something we achieve by using `Header.encode` first. Afterwards our data is ASCII so we never run into an odd double encoding situation. That said: There will be dragons (or so)

carltongibson commented:
Super. OK. So I guess that would be the force of my: 

> Is it worth a comment for future people looking at this, including ourselves... 

But I'll leave that there for now. 

(Hard isn't it. 😅)

felixxm commented:
@carltongibson @ewjoachim Thanks for reviews :+1: 

@apollo13 Thank for updates :+1: 

I added an extra check and tests for header injections in `sanitize_address()` (it's already protected in Django because we always go through `forbid_multi_line_headers()`, but it's worth adding in case someone or we will use it directly in the future).

I also added release notes and extra tests for preventing header injection.

carltongibson commented on docs/releases/3.0.9.txt:
```diff
@@ -11,3 +11,6 @@ Bugfixes
 * Allowed setting the ``SameSite`` cookie flag in
   :meth:`.HttpResponse.delete_cookie` (:ticket:`31790`).
+
+* Fixed sending emails crash on email addresses with display names longer than
+  75 chars on Python 3.6.11+, 3.7.8+, and 3.8.4+ (:ticket:`31784`).
```
```suggestion
* Fixed crash when sending emails to addresses with display names longer than
  75 chars on Python 3.6.11+, 3.7.8+, and 3.8.4+ (:ticket:`31784`).
```

carltongibson commented on docs/releases/2.2.15.txt:
```diff
@@ -11,3 +11,6 @@ Bugfixes
 * Allowed setting the ``SameSite`` cookie flag in
   :meth:`.HttpResponse.delete_cookie` (:ticket:`31790`).
+
+* Fixed sending emails crash on email addresses with display names longer than
+  75 chars on Python 3.6.11+, 3.7.8+, and 3.8.4+ (:ticket:`31784`).
```
```suggestion
* Fixed crash when sending emails to addresses with display names longer than
  75 chars on Python 3.6.11+, 3.7.8+, and 3.8.4+ (:ticket:`31784`).
```

felixxm commented on django/core/mail/message.py:
```diff
@@ -96,16 +96,24 @@ def sanitize_address(addr, encoding):
-    parsed_address = Address(nm, username=localpart, domain=domain)
-    return str(parsed_address)
+    parsed_address = Address(username=localpart, domain=domain)
+    return formataddr((nm, parsed_address.addr_spec), encoding)
```
`nm` is already encoded so it's not necessary to pass `encoding`. I will remove it.
